delphin/util.py

```diff
@@ -272,3 +272,43 @@ def peek(self, n=0, skip=None, drop=False):
             self._buffer_fill(n + 1)
             datum = buffer[n]
         return datum
+
+
+# modified from https://www.python.org/dev/peps/pep-0263/#defining-the-encoding
+_encoding_symbol_re = re.compile(
+    r'^[ \t\f]*.*?coding[:=][ \t]*([-_.a-zA-Z0-9]+)', re.IGNORECASE)
+
+import codecs, io
+
+def detect_encoding(filename, default_encoding='utf-8', comment_char=b';'):
+    encoding = None
+    with io.open(filename, 'rb') as fh:
+        line1 = next(fh)
+        re_match1 = _encoding_symbol_re.search(line1)
+        line1_comment = line1.contains(comment_char)
+        if line1_comment and re_match1:
+            encoding = _check_encoding(re_match1.group(1))
+            if not encoding and line1 != "":
```

Comment:
Why `and line1 != ""`? If the code reached this point, then `line1` at least has a comment character on it.

Also, if the file is opened in binary mode, any strings you compare to lines from the file should be byte strings (e.g., `b""`), otherwise they won't match in Python3:

```python3
>>> b'' == ''
False
```